Add testopt, a testing subpackage

vladimir-ch opened the pull request:
* Created testopt subpackage
* Copied objective functions from unconstrained_test.go
* Improved their documentation
* Most importantly, added tests that revealed mistakes in four functions. http://people.sc.fsu.edu/~jburkardt/m_src/test_opt/test_opt.html was used as a reference implementation
* Functions now return a list of their known minima

Modifying unconstrained_test.go to actually use these new, fairly reliably correct functions will come in another PR.

vladimir-ch commented:
PTAL, @btracey

btracey commented on testopt/testopt.go:
```diff
@@ -0,0 +1,132 @@
+	defaultGradTol = 1e-4
+)
+
+func testFunction(f Function, tests []funcTest, t *testing.T) {
```
In addition to (instead of?) wantGrad, maybe we should also test a couple of random points and compare with gonum/diff/fd.Gradient. For wantGrad, it's tempting to choose "simple" x locations like 0 and 1, which are the most likely to cause false positives. It's also a way to test the higher-dimensional functions where we have no interest in explicitly writing the gradient. fd.Gradient can be executed concurrently if we're worried about testing time.

btracey replied:
To be clear, I'm thinking of this as a second loop after the first range.

vladimir-ch replied:
Using fd.Gradient() and some random locations has crossed my mind. The locations I test now are the standard starting locations and the optimal locations.

I will try to add fd.Gradient() at random locations, but it might turn out as tricky to do in a general case. Some functions grow so fast away from a minimum that the numerical gradient can be very inaccurate. I'll see.

btracey replied:
I've found fd.Gradient works surprisingly well,  even for function values
up in the 10^11 range. You could also specify an  [][2]float64 with  min/max
On Feb 4, 2015 4:09 PM, "Vladimír Chalupecký" <notifications@github.com>
wrote:

> In testopt/testopt.go
> <https://github.com/gonum/optimize/pull/63#discussion_r24132771>:
>
> > +}
> > +
> > +type Gradient interface {
> > +	Df(x, grad []float64)
> > +}
> > +
> > +type FunctionGradient interface {
> > +	FDf(x, grad []float64) float64
> > +}
> > +
> > +const (
> > +	defaultTol     = 1e-8
> > +	defaultGradTol = 1e-4
> > +)
> > +
> > +func testFunction(f Function, tests []funcTest, t *testing.T) {
>
> Using fd.Gradient() and some random locations has crossed my mind. The
> locations I test now are the standard starting locations and the optimal
> locations.
>
> I will try to add fd.Gradient() at random locations, but it might turn out
> as tricky to do in a general case. Some functions grow so fast away from a
> minimum that the numerical gradient can be very inaccurate. I'll see.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/optimize/pull/63/files#r24132771>.
>

btracey commented:
Is the plan to move TestMinimize (hopefully renamed to TestLocal) and the other UnconstrainedTest lists into this package as well?

btracey commented on testopt/testopt.go:
```diff
@@ -0,0 +1,132 @@
+
+	if minima := f.Minima(); minima != nil {
+		for i := range minima {
+			if minima[i].X == nil {
```
I see that this case is used, but it's weird. You know what the minimum is, but you don't know where it is? Is there any utility in listing the minimum in that case?

vladimir-ch replied:
X can be nil in two cases:
1) the source paper lists the minimum value but not the location (possibly because it is not known analytically and listing the (inaccurate) numbers is too troublesome) and I haven't gotten around to run our optimizers to get it. Those few cases I will fix and provide also the location.
2) Some functions (from the implemented ones only two, if I am not mistaken) have minima on a line or plane. In that case I list at least the optimal value.

btracey replied:
It seems like we should always enforce x != nil. It doesn't help test anything, and it's somewhat confusing. The examples that fall into case 1) we can fix, and the examples that fall into case 2) we can return any of the values that lie along the line/plane. Other points returned can check that they are sufficiently close in function value and/or that the gradient is sufficiently small.

vladimir-ch replied:
3) Another case is when the function is variably dimensioned and the minimum is independent of the dimension (ExtendedPowelSingular, ExtendedRosenbrock, Trigonometric)

Re 1) The only case is the Watson function at dim = 12. Our minimizers find a minimum, but the function value does not agree with the one given in More, Garbow, Hillstrom. I will add a TODO.

btracey replied:
It's not true for ExtendedRosenbrock. There are multiple local minima in the higher dimensions. In any event, the result is the same as far as an optimizer is concerned -- the true test for finding a minima is finding a point with a sufficiently small gradient. The function value tests are only useful for confirming you have implemented the problem correctly.

With that in mind, if we have implemented the function properly and have found a point with zero gradient, then we have found an unlisted minimum.

vladimir-ch replied:
> It's not true for ExtendedRosenbrock. There are multiple local minima in the higher dimensions.

True. I was thinking about a global minimum, the three functions I mentioned have a global minimum with a value of 0, for ExtendedRosenbrock its location is (1, ..., 1).

> ... then we have found an unlisted minimum.

Or a false minimum where the gradient is very small and finite precision does not allow further progress. It is possible that this is what is happening with the Watson function.

I have no problem with enforcing x != nil, I will modify the code accordingly.

btracey commented on testopt/testopt.go:
```diff
@@ -0,0 +1,132 @@
+				grad := make([]float64, len(test.wantGrad))
+				F := fFunctionGradient.FDf(test.x, grad)
+
+				if !floats.EqualWithinAbs(F, test.wantF, defaultTol) {
```
Personally, I prefer using just math.Abs() > defaultTol. Seems simpler than going to the floats function.

vladimir-ch replied:
OK, I will change it.

btracey commented:
What are your thoughts on exporting the testFunction to allow an easy check for outsiders with functions? It may be useful, but on the other hand maybe we don't want to guarantee anything (and maybe this isn't robust enough for complicated functions).

I think it's worth including a doc comment for this package that says something along the lines of "This is a testing yadda yadda. We encourage outside contributions of additional test functions that exhibit properties not already covered in the testing suite." (maybe something extra about 'or with significance', due to use in benchmark cases or whatever)

btracey commented:
Thanks a ton for putting together so many test functions. It's really important for the package.

vladimir-ch commented:
Moving TestLocal() to testopt: I didn't consider it, but perhaps it should be where the test cases are. If we move them and export in testopt, then TestLocal() could be there as well.

Exporting testFunction(): good idea, especially if we compare the computed gradient with fd.Gradient(), that would be very useful. I am for exporting it (after a proper revision of its signature and the tolerance checks inside it). I will add an issue.

Right now I am driven by the curiosity about why the optimization from a nearly optimum sometimes fails. I wonder if some of the failures were caused by mistakes in the test functions, when the function values were not consistent with the gradient.

I will try to add some doc comments as you suggest (too bad you wrote yadda yadda, otherwise I would just copy/paste it :-)

btracey commented:
If we move TestLocal() etc. here and export TestFunction(), then there is a purpose clash. Does this package test optimizers or does it test objective functions? I propose that we rename this package to "functions". Then a user could use functions.Test() on their own, or take functions.Beale{} as an example function. For testing optimizers we have a second package that is testopt, which could use the functions provided in functions (or we could choose to keep the testing suite private). This way seems to keep reasonably sized packages with one idea in mind.

With this change, I think the requirement on having Minima should be dropped, instead replaced by a Minimimumer() interface that is type asserted. Most user functions will have no idea on what/where their minima are, and implementing Minimumer is not required for using Local.

vladimir-ch commented:
> If we move TestLocal() etc. here and export TestFunction(), then there is a purpose clash. Does this package test optimizers or does it test objective functions?

True.

> I propose that we rename this package to "functions". Then a user could use functions.Test() on their own, or take functions.Beale{} as an example function.

Very much agreed.

> For testing optimizers we have a second package that is testopt, which could use the functions provided in functions (or we could choose to keep the testing suite private). This way seems to keep reasonably sized packages with one idea in mind.

At the moment I would keep it private until the optimizers and tests settle down.

> With this change, I think the requirement on having Minima should be dropped, instead replaced by a Minimimumer() interface that is type asserted. Most user functions will have no idea on what/where their minima are, and implementing Minimumer is not required for using Local.

OK

vladimir-ch commented:
* Added finite difference gradient evaluation which revealed one more, hopefully last, error in a function (HelicalValley).
* Added Minimumer interface
* FuncTest.X must not be nil
* Did _not_ add exported Test() function, need to think about its arguments, etc first.
* Did _not_ add evaluation at random points, need to think how to choose those points first.

PTAL, @btracey

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+// algorithms.
+//
+// We encourage outside contributions of additional test functions that exhibit
+// properties not already covered in the testing suite or with significance due
```
or that have significance due to prior use as benchmark cases.

vladimir-ch replied:
Thanks. Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+)
+
+// Function represents an objective function.
+type Function interface {
```
These interfaces can all be private until we decide on the proper API to export.

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates whether the minimum is local or global.
```
Global indicates if the location is a global minimum.

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+	Global bool
+}
+
+type FuncTest struct {
```
I think this can be private too.

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+
+const (
+	defaultTol     = 1e-8
+	defaultGradTol = 1e-5
```
There should be two gradient tolerances. One for the specifed gradient value, which is tight (1e-14 ish), and one for finite difference which is loose (1e-6).

I think the function tolerance should be increased to 1e-14. We're comparing to a supposedly known value, so we just have to prevent against floating point noise.

vladimir-ch replied:
I added a second gradient tolerance, that was a good idea. However, the suggested tolerances are too tight. For function values I use 1e-12 (smaller values show floating point noise in ExtendedRosenbrock). For Df() gradients I use 1e-9. Some minima are simply not known accurately enough. I disabled the test for BrownAndDennis because otherwise the value would have to be larger than 1e-6.  And for the numerical gradient the tolerance is 1e-5,  BrownBadlyScaled is the culprit (with BrownAndDennis the tolerance would have to be 1e-4).

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+)
+
+func testFunction(f Function, tests []FuncTest, t *testing.T) {
+	// Check that the function can evaluate itself (and its gradient) correctly.
```
this should use godoc style and be before the declaration.

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+type FuncTest struct {
+	X []float64
+
+	WantF    float64
```
Do these need the "Want" prefixes? It's clear from context, isn't it?

vladimir-ch replied:
Removed Want prefix.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+
+		// Evaluate the finite difference gradient.
+		fdGrad := make([]float64, len(test.WantGrad))
+		fd.Gradient(fdGrad, func(x []float64) float64 {
```
You can just pass f.F directly

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+		}
+
+		// Evaluate the finite difference gradient.
+		fdGrad := make([]float64, len(test.WantGrad))
```
You can do 

fdGrad := fd.Gradient(nil, f.F, test.X, nil)

vladimir-ch replied:
Great. I should have paid more attention to the documentation.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,187 @@
+		return
+	}
+
+	// Check that the listed minima are correct.
```
Can we just append the minima to the fd test? Seems like it would reduce all this code duplication.

vladimir-ch replied:
Good idea. Done.

vladimir-ch commented:
I have added some local minima to ExtendedRosenbrock and I have addressed all the comments. PTAL @btracey

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,157 @@
+	"github.com/gonum/floats"
+)
+
+// Function represents an objective function.
```
lowercase comment

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,157 @@
+	Gradient []float64
+}
+
+// TODO: Decide and implement an exported testing function:
```
Please have username with TODO

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -0,0 +1,157 @@
+// correctly.
+func testFunction(f function, ftests []funcTest, t *testing.T) {
+	// Make a copy of tests because we may append to the slice.
+	// TODO: is this necessary?
```
Just take off this TODO. A better solution isn't obvious to me, and I'm not worried about the memory allocation.

vladimir-ch replied:
Done.

btracey commented:
I don't like your file names. I would have assumed that the functions themselves live in "functions.go", and I have no idea what mgh refers to.

I would suggest renaming mgh.go to functions.go and rename functions.go to validate.go

LGTM with the documentation and file name fixes.

vladimir-ch commented:
Fixed the documentation and the file names. Thanks for your comments, @btracey .
